SO-3294: Support word highlighting in validation issues

AAAlinaaa opened the pull request:
This PR adds support for word highlighting in the validation issues API.

AAAlinaaa committed:
SO-3294: Handle different return types by the validation rule ...

... evaluators, include highlighting details returned by some rules in
the validation issue details

cmark commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/validation/ValidationIssueDetails.java:
```diff
@@ -0,0 +1,25 @@
+package com.b2international.snowowl.core.validation;
```
⚠️ Missing license!

cmark commented on core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/validation/ValidateRequest.java:
```diff
@@ -254,11 +264,19 @@ void setRuleParameters(Map<String, Object> ruleParameters) {
+		public Collection<ValidationIssueDetails> issueDetails;
+		
+		@SuppressWarnings("unchecked")
+		public <T> IssuesToPersist(String ruleId, Collection<T> evaluationResult) {
```
This `<T>` type parameter is completely unnecessary. Just use the `?` wildcard type instead.

cmark commented on core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/validation/ValidateRequest.java:
```diff
@@ -254,11 +264,19 @@ void setRuleParameters(Map<String, Object> ruleParameters) {
+			if (evaluationResult.iterator().hasNext() && evaluationResult.iterator().next() instanceof ValidationIssueDetails) {
+				this.issueDetails = (Collection<ValidationIssueDetails>) evaluationResult;
+			} else {
+				this.issueDetails = evaluationResult.parallelStream()
```
You should not use `parallelStream` to process a huge list like this, none of the downstream consumers do anything that would indicate the necessity of parallel execution.

cmark commented on core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/validation/eval/GroovyScriptValidationRuleEvaluator.java:
```diff
@@ -40,7 +39,7 @@ public GroovyScriptValidationRuleEvaluator(Path validationResourcesDirectory) {
 	
 	@Override
-	public List<ComponentIdentifier> eval(BranchContext context, ValidationRule rule, Map<String, Object> filterParams) throws Exception {
+	public <T> List<T> eval(BranchContext context, ValidationRule rule, Map<String, Object> filterParams) throws Exception {
```
Use wildcard `?` here as well.

cmark commented on core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/validation/ValidateRequest.java:
```diff
@@ -202,10 +205,17 @@ private ValidationResult doValidate(BranchContext context, Writer index) throws
 						final Collection<ValidationIssue> issues = issuesToExtendWithDetailsByToolingId.removeAll(toolingId);
 						extensions.extendIssues(context, issues);
+						
+						/*ValidationIssueDetailExtensionProvider
```
Commented code is bad code, remove it!

cmark commented on core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/validation/eval/ValidationRuleEvaluator.java:
```diff
@@ -76,12 +75,13 @@ public static ValidationRuleEvaluator get(String type) {
 	 * @throws Exception
 	 */
-	List<ComponentIdentifier> eval(BranchContext context, ValidationRule rule, Map<String, Object> params) throws Exception;
+	<T> List<T> eval(BranchContext context, ValidationRule rule, Map<String, Object> params) throws Exception;
```
Same here, use wildcard `?` type.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/validation/SnomedQueryValidationRuleEvaluator.java:
```diff
@@ -79,7 +79,7 @@
 
 	@Override
-	public List<ComponentIdentifier> eval(BranchContext context, ValidationRule rule, Map<String, Object> params) throws Exception {
+	public <T> List<T> eval(BranchContext context, ValidationRule rule, Map<String, Object> params) throws Exception {
```
And here!

cmark commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/validation/ValidationIssueDetails.java:
```diff
@@ -0,0 +1,25 @@
+		this.affectedComponentId = affectedComponentId;
+	}
+	
+	public ValidationIssueDetails(List<Entry<Integer, Integer>> detailEntries, ComponentIdentifier affectedComponentId) {
```
A `List<Entry<Integer, Integer>>` is essentially a `Map` if the list should not contain any duplicate entry, which in this case is true, I think. Also, rename the `detailEntries` field to something meaningful.
Please refactor.

AAAlinaaa committed:
SO-3294: Remove commented out code, unnecessary type parameters, use...

... of paralllelstreams, add missing license header & since tag to
ValidationIssueDetails

AAAlinaaa committed:
SO-3294: Use new pojo to store styling information

AAAlinaaa committed:
SO-3294: Add license header, since tag

cmark commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/validation/StylingDetail.java:
```diff
@@ -0,0 +1,43 @@
+	private final int index;
+	private final int length;
+	
+	public StylingDetail(
```
Just in case, annotate the constructor with `JsonCreator` annotation.

cmark commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/validation/StylingDetail.java:
```diff
@@ -0,0 +1,43 @@
+/**
+ * @since 6.16
+ */
+public class StylingDetail {
```
Make this class `final`.

cmark commented on core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/validation/ValidateRequest.java:
```diff
@@ -254,11 +259,19 @@ void setRuleParameters(Map<String, Object> ruleParameters) {
-		public final Collection<ComponentIdentifier> affectedComponentIds;
-
-		public IssuesToPersist(String ruleId, Collection<ComponentIdentifier> affectedComponentIds) {
+		public Collection<ValidationIssueDetails> issueDetails;
```
This field should be final.

cmark commented on core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/validation/ValidateRequest.java:
```diff
@@ -105,8 +106,8 @@ private ValidationResult doValidate(BranchContext context, Writer index) throws
 						LOG.info("Executing rule '{}'...", rule.getId());
-						final List<ComponentIdentifier> componentIdentifiers = evaluator.eval(context, rule, ruleParameters);
-						issuesToPersistQueue.offer(new IssuesToPersist(rule.getId(), componentIdentifiers));
+						final List evaluationResponse = evaluator.eval(context, rule, ruleParameters);
```
Use wildcard type `List<?>` here as well.

AAAlinaaa committed:
SO-3294: Make Styling Detail final, add @JsoneCreator annotation, ...

... use wild card type to list evaluation results in ValidateRequest,
make validation issue details field final in IssuesToPersist

cmark reviewed (approved):
LGTM!
